File: 0016. 3Sum Closest/0016.java
// time complexity: O(n^2)
// space complexity: O(1)
class Solution {
    // returns the sum of three integers in nums that is closest to the target.
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);

        int closest = Integer.MAX_VALUE;
        for(int i=0; i<nums.length-2; i++){
            int left = i+1;
            int right = nums.length - 1;

            // for every i
            while(left<right){
                int sum = nums[i] + nums[left] + nums[right];
                // update the closest
                if(Math.abs(closest-target) > Math.abs(sum-target)){
                    closest = sum;
                }
                // move the pointers
                if(sum>target) right--;
                else left++;
            }
        }

        return closest;
    }
}

File: 0016. 3Sum Closest/0016.py
# time complexity: O(n^2)
# space complexity: O(1)
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        closest = sum(nums[:3]) # initialize using sum of first three numbers

        for i in range(len(nums) - 2):
            left = i+1
            right = len(nums)-1

            # use two pointers to find best pair for nums[i]
            while left < right:
                current_sum = nums[i] + nums[left] + nums[right]

                # update closest
                if abs(target-current_sum) < abs(target-closest):
                    closest = current_sum
                
                # move pointers
                if current_sum < target:
                    left += 1
                else:
                    right -= 1
        
        return closest
                
